mlos_core/mlos_core/optimizers/observations.py

```diff
@@ -0,0 +1,252 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""Contains classes for observations and suggestions."""
+
+from __future__ import annotations
+
+from typing import Any, Iterator, Optional
+
+import pandas as pd
+
+
+def compare_optional_dataframe(
+    left: Optional[pd.DataFrame], right: Optional[pd.DataFrame]
+) -> bool:
+    """
+    Compare DataFrames that may also be None.
+
+    Parameters
+    ----------
+    left : Optional[pd.DataFrame]
+        The left DataFrame to compare
+    right : Optional[pd.DataFrame]
+        The right DataFrame to compare
+
+    Returns
+    -------
+    bool
+        Compare the equality of two Optional[pd.DataFrame] objects
+    """
+    if left is None and right is not None:
+        return False
+    if left is not None and right is None:
+        return False
+    elif left is not None and right is not None:
+        if not left.equals(right):
+            return False
+    return True
```

Comment:
I think this can be simplified a bit:

```python
if left is not None and right is not None and left.equals(right):
  return True
return False
```